orttraining/orttraining/training_ops/cuda/communication/nccl_service.cc

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef USE_NCCL
+
+#include "core/common/common.h"
+#include "core/profile/context.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "orttraining/core/framework/mpi_context.h"
+#include "orttraining/training_ops/cuda/communication/nccl_service.h"
+#include <iostream>
+#include <nccl.h>
+
+namespace onnxruntime {
+namespace cuda {
+
+bool NcclTask::Compare(const NcclTask& other) const {
+  if (type != other.type) {
+    return false;
+  }
+  if (peers.size() != other.peers.size()) {
+    return false;
+  }
+  for (size_t i = 0; i < peers.size(); ++i) {
+    if (peers[i] != other.peers[i]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void NcclTask::ResetTask() {
+  ptr = nullptr;
+  size = 0;
+  is_enqueued = false;
+  is_finished = false;
+}
+
+void NcclTaskGroup::PlanTask(
+    const NcclTask::Type type,
+    const std::vector<int> peers) {
+  batch.push_back({type, peers, nullptr, 0, false, false, ""});
+};
+
+const NcclTask* NcclTaskGroup::EqueueTask(
+    const NcclTask::Type type,
+    const std::vector<int> peers,
+    void* ptr,
+    const size_t size,
+    const std::string info) {
+  NcclTask scheduled_task;
+  scheduled_task.type = type;
+  scheduled_task.peers = peers;
+
+  for (auto& task : batch) {
+    if (!task.Compare(scheduled_task)) {
+      continue;
+    }
+
+    // We cannot enqueue the same task.
+    ORT_ENFORCE(!task.is_finished);
+    // We cannot enqueue the same task.
+    ORT_ENFORCE(!task.is_enqueued);
+
+    task.ptr = ptr;
+    task.size = size;
+    task.is_enqueued = true;
+    task.info = info;
+    return &task;
+  }
+
+  return nullptr;
+};
+
+bool NcclTaskGroup::IsAllTasksEqueued() const {
+  return std::all_of(
+      batch.begin(), batch.end(), [&](const NcclTask& task) {
+        return task.is_enqueued;
+      });
+};
+
+bool NcclTaskGroup::IsAllTasksFinished() const {
+  return std::all_of(
+      batch.begin(), batch.end(), [&](const NcclTask& task) {
+        return task.is_finished;
+      });
+};
+
+void NcclTaskGroup::ResetAllTasks() {
+  for (auto& task : batch) {
+    task.ptr = nullptr;
+    task.size = 0;
+    task.is_enqueued = false;
+    task.is_finished = false;
+  }
+};
+
+void NcclService::PlanStart() {
+  ORT_ENFORCE(!is_planned_, "Communication plan cannot be changed after calling PlanEnd.");
+};
+
+void NcclService::PlanEnd() {
+  is_planned_ = true;
+};
+
+void NcclService::PlanNewGroupStart() {
+  group_status_.push_back(true);
+  schedule_.push_back(NcclTaskGroup());
+};
+
+void NcclService::PlanNewGroupEnd() {
+  group_status_.back() = false;
+};
+
+void NcclService::PlanSend(const int dst) {
+  ORT_ENFORCE(group_status_.back(), "Last communication group can not be changed after call PlanEndNewGroup.");
```

Comment:
>ORT_ENFORCE(group_status_.back(), "Last communication group can not be changed after call PlanEndNewGroup."); [](start = 2, length = 109)

consider `ORT_ENFORCE(group_status_.size() > 0 && group_status_.back(), "Last communication group doesn't exist or attempt to be changed after call PlanEndNewGroup.");`

Reply:
similar for `NcclService::PlanRecv`

---
In reply to: [485985106](https://github.com/microsoft/onnxruntime/pull/5054#discussion_r485985106) [](ancestors = 485985106)